modules/oncodrivefml/1.0/oncodrivefml.smk

```diff
@@ -0,0 +1,252 @@
+    log:
+        stdout = CFG["logs"]["oncodrivefml"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.stdout.log",
+        stderr = CFG["logs"]["oncodrivefml"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.stderr.log"
+    params:
+        local_path = CFG["reference_files_directory"],
+        build = lambda w: ONCODRIVE_BUILD_DICT[w.genome_build],
+        command_line_options = CFG["options"]["fml_options"] if CFG["options"]["fml_options"] is not None else ""
+    threads:
+        CFG["threads"]["fml"]
+    resources:
+        **CFG["resources"]["fml"]
+    conda: CFG["conda_envs"]["fml"]
+    shell:
+        op.as_one_line("""
+        export BGDATA_LOCAL={params.local_path} &&
+        export BGDATA_OFFLINE=TRUE &&
+        oncodrivefml
+        -i {input.maf}
+        -o "$(dirname $(realpath {output.tsv}))"
+        -e {input.region}
+        -c {input.config}
+        --cores {threads}
+        {params.command_line_options}
+        > {log.stdout} 2> {log.stderr}
+        """)
+
+rule _oncodrivefml_out:
+    input:
+        tsv = str(rules._oncodrivefml_run.output.tsv),
+        png = str(rules._oncodrivefml_run.output.png),
+        html = str(rules._oncodrivefml_run.output.html)
+    output:
+        tsv = CFG["dirs"]["outputs"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.tsv",
+        png = CFG["dirs"]["outputs"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.png",
+        html = CFG["dirs"]["outputs"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.html"
+    run:
+        op.relative_symlink(input.tsv, output.tsv, in_module=True)
+        op.relative_symlink(input.png, output.png, in_module=True)
+        op.relative_symlink(input.html, output.html, in_module=True)
+
+def _get_oncodrivefml_outputs(wildcards):
+    CFG = config["lcr-modules"]["oncodrivefml"]
+    checkpoint_output = os.path.dirname(str(checkpoints._oncodrivefml_prep_input.get(**wildcards).output[0]))
+    SUMS, = glob_wildcards(checkpoint_output+"/{md5sum}.maf.content")
+
+    return expand(
+        [
+            CFG["dirs"]["outputs"] + "{{genome_build}}/{{sample_set}}--{{launch_date}}/{md5sum}/{{region}}/oncodrivefml.tsv",
+            CFG["dirs"]["outputs"] + "{{genome_build}}/{{sample_set}}--{{launch_date}}/{md5sum}/{{region}}/oncodrivefml.png",
+            CFG["dirs"]["outputs"] + "{{genome_build}}/{{sample_set}}--{{launch_date}}/{md5sum}/{{region}}/oncodrivefml.html"
+        ],
+        md5sum = SUMS
+    )
+
+rule _oncodrivefml_aggregate:
+    input:
+        _get_oncodrivefml_outputs
+    output:
+        aggregate = CFG["dirs"]["outputs"] + "{genome_build}/{sample_set}--{launch_date}/aggregate/{sample_set}--{launch_date}--{region}.done"
+    shell:
+        "touch {output.aggregate}"
```

Comment:
Usually there is a symlinking rule at the end, but here the dummy file? What is this rule intended to do here?

Reply:
`rule _oncodrivefml_out:
    input:
        tsv = str(rules._oncodrivefml_run.output.tsv),
        png = str(rules._oncodrivefml_run.output.png),
        html = str(rules._oncodrivefml_run.output.html)
    output:
        tsv = CFG["dirs"]["outputs"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.tsv.gz",
        png = CFG["dirs"]["outputs"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.png",
        html = CFG["dirs"]["outputs"] + "{genome_build}/{sample_set}--{launch_date}/{md5sum}/{region}/oncodrivefml.html"
    run:
        op.relative_symlink(input.tsv, output.tsv, in_module=True)
        op.relative_symlink(input.png, output.png, in_module=True)
        op.relative_symlink(input.html, output.html, in_module=True)

def _get_oncodrivefml_outputs(wildcards):
    CFG = config["lcr-modules"]["oncodrivefml"]
    checkpoint_output = os.path.dirname(str(checkpoints._oncodrivefml_prep_input.get(**wildcards).output[0]))
    SUMS, = glob_wildcards(checkpoint_output+"/{md5sum}.maf.content")

    return expand(
        [
            CFG["dirs"]["outputs"] + "{{genome_build}}/{{sample_set}}--{{launch_date}}/{md5sum}/{{region}}/oncodrivefml.tsv.gz",
            CFG["dirs"]["outputs"] + "{{genome_build}}/{{sample_set}}--{{launch_date}}/{md5sum}/{{region}}/oncodrivefml.png",
            CFG["dirs"]["outputs"] + "{{genome_build}}/{{sample_set}}--{{launch_date}}/{md5sum}/{{region}}/oncodrivefml.html"
        ],
        md5sum = SUMS
    )`
    
The last rule looks for the final output files symlinked to 99-outputs by expanding the md5sum wildcard. This rule is a dummy file because it requires an input function to expand the md5sum values so it can't be in rule all I believe, because rule all expands based on "genome_build", "sample_set", "launch_date", and "regions" but to expand the md5sum wildcard it requires checking the outputs of the checkpoint rule